### ISSUE TITLE
Audit ES graph controller placement at the host boundary

### ISSUE THREAD

**Issue by Ahmadfauzi34** [Ahmadfauzi34/ai_agent_builder#177]
Follow-up after #174 proved the packaged host lifecycle `ask -> apply -> graph.run(s) -> objective -> tell`, and #176/#175 showed the existing stateful ProgramBundle is sufficient for checkpoint/replay semantics.

The next question is architectural placement, not a request to add a facade immediately:

> Should ES + graph orchestration become a Rust/WASM controller/objective abstraction, or remain a host-level policy composed from the existing proof-bounded primitives?

## Current evidence

- `EsOptimizer` is intentionally objective-agnostic and exposes strict `ask -> tell` lifecycle semantics.
- `src/es/objective.rs` contains an internal Rust `Objective::fitness(&self, params)` proof-of-life abstraction, but it is not a general WASM objective callback surface.
- graph candidate application requires mutable `LayerRegistry`, while `CompiledGraph.run` consumes the registry immutably after application.
- #174 measured graph-parameter apply overhead and found it strongly amortized under realistic multi-forward objectives; no reusable binding/cache was justified.
- #176 proved best learned state can already be promoted/replayed through `ProgramBundle(include_state=true)`; no new checkpoint identity is needed.
- the packaged Node host adapter is deliberately thin and other hosts (Rust/Python/Go) are planned, so objective/controller policy should not accidentally become Node-only or Rust-only without workload evidence.

## Research goal

Prove whether the host boundary can express at least two materially different objective policies using exactly the same optimizer/graph/binding/checkpoint primitives, without host-side parameter reordering or hidden mutable state:

1. deterministic regression MSE over multiple graph runs;
2. a composite objective (for example MSE plus an explicit finite parameter/output penalty or held-out component) whose policy lives only at the host layer.

For both workloads prove:

```text
graphParameterLayout.total_len
    -> EsOptimizer.ask
    -> canonical setGraphParametersFlat
    -> CompiledGraph.run(s)
    -> host objective policy
    -> finite fitness vector
    -> EsOptimizer.tell
    -> best
    -> ProgramBundle(include_state=true)
    -> fresh import/replay
```

### KEEP_HOST_ORCHESTRATION
Choose this if the existing primitives support both objectives cleanly and the host controller is only policy/glue. In that case:
- do not change `Objective` mutability;
- do not add a Rust `GraphController` or graph-owning optimizer facade;
- do not add a JS-callback bridge into Rust;
- keep checkpoint artifact separate from authorization/proof authority.

### PROPOSE_CORE_CONTROLLER
Only consider a core controller if the research finds a concrete invariant that cannot be enforced compositionally at the host boundary, such as an unavoidable partial-mutation window, ambiguous candidate ordering, lifecycle race, or cross-host semantic mismatch.

Any proposal must name that missing invariant and prove why a new ownership/controller abstraction closes it.

## Non-goals

- no Math Program changes;
- no autodiff/backprop;
- no new optimizer algorithm;
- no public reusable graph-parameter cache;
- no new checkpoint/state identity;
- no byte-SHA promotion semantics for ProgramBundle;
- no Resolution/Authorization coupling;
- no production controller until the placement question is answered by a packaged workload.

**Comment by Ahmadfauzi34** [comment]
Packaged-artifact research completed directly against the exact PR #176 artifact (already digest-verified), without rebuilding source.

Two materially different objective policies were run with the same runtime primitives and canonical candidate ordering:

### Objective A — pure regression MSE
- initial loss: `3.57666015625`
- final loss: `0.0014337145141809504`
- best: `[1.5035406351089478, -0.7666816711425781, 0.28625044226646423]`
- parameter norm: `1.7118336782490116`
- stateful ProgramBundle fresh-import replay: PASS

### Objective B — host policy `MSE + 0.15 * mean(params^2)`
- initial loss: `3.57666015625`
- final composite loss: `0.13842774792599954`
- final base MSE: `0.003974352849024712`
- final regularization penalty: `0.13445339507697485`
- best: `[1.4583685398101807, -0.7082248330116272, 0.2462654858827591]`
- parameter norm: `1.6398377668353346`
- stateful ProgramBundle fresh-import replay: PASS

The policy effect is observable: best candidates differ and the regularized policy produces a smaller parameter norm, while `EsOptimizer`, `CompiledGraph`, `setGraphParametersFlat`, graph-parameter layout/order, and checkpoint semantics remain unchanged.

No host-side parameter reordering was used: candidate slices came directly from `EsOptimizer.ask()` using `graphParameterLayout.total_len`, then were applied through canonical `setGraphParametersFlat` before graph execution.

## Decision

`KEEP_HOST_ORCHESTRATION`

No concrete invariant gap was found that requires a Rust/WASM graph-owning controller, mutable `Objective` redesign, JS-callback bridge into Rust, or optimizer/graph ownership fusion.

The current architecture should remain:

```text
Core/WASM: strict primitives + proof boundaries
  EsOptimizer.ask/tell
  GraphParameterBinding apply/read
  CompiledGraph.run
  ProgramBundle checkpoint/replay

Host policy:
  objective composition
  dataset traversal
  penalties / held-out policy
  training loop orchestration
```

The old comment in `src/es/objective.rs` about a future graph-running Objective should therefore be treated as historical intent, not as a requirement. A core controller should only be reopened if a future workload demonstrates an invariant that host composition cannot enforce.
